### PR TITLE
Improvements and fixes

### PR DESCRIPTION
**Pull request by brunomvsouza**
- Improving error messages
- Fixing errors due to absense of a default nullable secondsBehindMaster and openConnections
- Improvements in test cases
- Fixing *int sort (https://play.golang.org/p/50bVfo2USy)

### CODE REVIEW

File: balancer.go [hunk 1]
```diff
@@ -12,6 +12,12 @@ type bySecondsBehindMaster Servers
 func (a bySecondsBehindMaster) Len() int      { return len(a) }
 func (a bySecondsBehindMaster) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
 func (a bySecondsBehindMaster) Less(i, j int) bool {
+	if a[i].health.secondsBehindMaster == nil && a[j].health.secondsBehindMaster != nil {
```
Comment (brunomvsouza) [inline balancer.go]: Handling `nil` shifting it to the end of the slice.

File: balancer.go [hunk 2]
```diff
@@ -20,7 +26,13 @@ type byOpenConnections Servers
 func (a byOpenConnections) Len() int      { return len(a) }
 func (a byOpenConnections) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
 func (a byOpenConnections) Less(i, j int) bool {
-	return a[i].health.openConnections < a[j].health.openConnections
+	if a[i].health.openConnections == nil && a[j].health.openConnections != nil {
```
Comment (brunomvsouza) [inline balancer.go]: Handling `nil` shifting it to the end of the slice.

File: balancer.go [hunk 3]
```diff
@@ -86,14 +95,17 @@ func (b *Balancer) PickServer() *Server {
 
 // New creates a new instance of Balancer
 func New(config *Config) *Balancer {
+	// Minimum check interval
+	if config.CheckInterval == 0 {
```
Comment (brunomvsouza) [inline balancer.go]: Moving one time configuration to the `struct` initialization.

File: balancer_test.go [hunk 4]
```diff
@@ -3,337 +3,231 @@ package balancer
...
 	"errors"
 	"testing"
-	"time"
 
-	"github.com/pborman/uuid"
 	. "github.com/smartystreets/goconvey/convey"
 )
 
+var (
+	ServerDownDueToMySQLConnection, ServerDownDueToMySQLSlaveStatus, ServerDownDueToMySQLThreadStatus         *Server
```
Comment (brunomvsouza) [inline balancer_test.go]: Reusable structs to make test cases easier to follow.

File: health.go [hunk 5]
```diff
@@ -6,36 +6,22 @@ import (
 )
 
 type ServerHealth struct {
-	up                  bool
-	err                 error
+	sync.Mutex
```
Comment (brunomvsouza) [inline health.go]: Embedding `sync.Mutex`.

File: server.go [hunk 6]
```diff
@@ -93,37 +95,65 @@ func (s *Server) rawQuery(query string, logger Logger) (map[string]string, error
...
 	if err != nil {
-		s.health.setDown(err)
+		s.health.setDown(
+			fmt.Errorf("unexpected value for Threads_connected returned from MySQL:  %s", err),
+			secondsBehindMaster, openConnections,
+		)
 		return
 	}
 
-	s.health.mutex.Lock()
```
Comment (brunomvsouza) [inline server.go]: Using only `health.setUP()` and `health.setDown()` to control `health`'s state to avoid race condition.

File: server.go [hunk 7]
```diff
@@ -93,37 +95,65 @@ func (s *Server) rawQuery(query string, logger Logger) (map[string]string, error
...
-			if v, err := strconv.Atoi(secondsBehindMaster); err == nil {
-				s.health.mutex.Lock()
-				s.health.secondsBehindMaster = &v
-				s.health.mutex.Unlock()
-			}
-		}
+	slaveStatusResult, err := s.rawQuery("SHOW SLAVE STATUS", logger)
+	if err != nil {
+		s.health.setDown(
+			fmt.Errorf("failed acquiring MySQL slave status: %s", err),
```
Comment (brunomvsouza) [inline server.go]: More descriptive errors.

**Comment by cicerocomp** [comment]
LGTM

**Pull request merged**